web/src/lib/actions/input.ts

```diff
@@ -0,0 +1,229 @@
+  handler: (event: KeyboardEvent) => void,
+): (event: KeyboardEvent) => void {
+  const formatCombo = (combo: KeyCombo) => {
+    const parts = [];
+    if (combo.ctrl) {
+      parts.push('Ctrl');
+    }
+    if (combo.alt) {
+      parts.push('Alt');
+    }
+    if (combo.shift) {
+      parts.push('Shift');
+    }
+    if (combo.meta) {
+      parts.push('Meta');
+    }
+    parts.push(combo.key);
+    return parts.join('+');
+  };
+
+  for (const shortcut of shortcuts) {
+    for (const reserved of ALL_RESERVED_SHORTCUTS) {
+      // Check if shortcuts match (comparing all properties)
+      if (
+        shortcut.key.toLowerCase() === reserved.key.toLowerCase() &&
+        !!shortcut.ctrl === !!reserved.ctrl &&
+        !!shortcut.alt === !!reserved.alt &&
+        !!shortcut.shift === !!reserved.shift &&
+        !!shortcut.meta === !!reserved.meta
+      ) {
+        console.error(
+          `[Keyboard Shortcut Warning] Attempting to register reserved shortcut: ${formatCombo(shortcut)}. ` +
+            `This shortcut is reserved by the OS or browser and may not work as expected.`,
+        );
+        return () => void 0;
+      }
+    }
+  }
+  return handler;
+}
+
+export const keyDownListenerFactory = (
+  isActiveFactory: () => () => boolean,
+  options: { ignoreInputFields?: boolean },
+  shortcuts: KeyCombo[],
+  callback: ShortcutCallback,
+) =>
+  checkForReservedShortcuts(shortcuts, (event: KeyboardEvent) => {
+    const isActive = isActiveFactory();
+    if (!isActive() || !isKeyboardEvent(event) || ((options.ignoreInputFields ?? true) && shouldIgnoreEvent(event))) {
+      return;
+    }
+
+    for (const currentShortcut of shortcuts) {
+      // pressing 'shift' will cause keyEvents to use capital key - adjust shortcut key to be capital to match
+      const matchingKey = currentShortcut.shift ? currentShortcut.key.toUpperCase() : currentShortcut.key;
+
+      // On mac, pressing 'alt+<somekey>' transforms the key to a special character
+      // but the code property has the physical key. If the code starts with Key/Digit
+      // extract the key from the code to consistently process alt keys on all platforms.
+      let baseKey = event.key;
+      const code = event.code;
+      if (code.startsWith('Key')) {
+        baseKey = code.slice(3).toLowerCase();
+      } else if (code.startsWith('Digit')) {
+        baseKey = code.slice(5);
+      }
+
+      if (
+        baseKey !== matchingKey ||
+        !!currentShortcut.ctrl !== event.ctrlKey ||
+        !!currentShortcut.alt !== event.altKey ||
+        !!currentShortcut.shift !== event.shiftKey
+      ) {
+        continue;
+      }
+
+      const result = callback(event);
+      if (isPromise(result) || result === false) {
+        // An event handler must be syncronous to call preventDefault
+        // if a handler is a promise then it can't rely on the automatic
+        // preventDefault() behavior, and must manually do that itself.
+        return;
+      }
```

Comment:
What's the harm in adding preventDefault here anyways? If it doesn't have any effect I don't see an issue with just removing that special-casing